feat(common): Add an 'outlet' injector option for ngTemplateOutlet

mmalerba opened the pull request:
[fix(core): correctly handle SkipSelf when resolving from embedded view injector](https://github.com/angular/angular/pull/55389/commits/2d25f668b47de9dc231fe4b6c048b0cd2eba4289) 

The `SkipSelf` flag is intended for the current injection context. When delegating to an embedded view injector, we are traversing to a parent/fallback scope, so the 'Self' has already been skipped.

[feat(common): add an 'outlet' injector option for ngTemplateOutlet](https://github.com/angular/angular/pull/55389/commits/479c145cc08931cdbb6ede3cccbf9541371587ae) 

Adds an option (`ngTemplateOutletInjector="outlet"`) that instructs the ngTemplateOutlet to inherit its injector from the outlet's place in the instantiated DOM.

JeanMeche commented:
Question: could this be a fix for #54097 ?

mmalerba commented:
Hmm I don't think so, the template outlet in that example looks like its already manually doing something pretty similar to what this would do. I tried adding an extra `Injector.create` to that example and it still had the same issue, so I think it would have the same issue with `"inherit"`. This change was motivated by making it easier to build recursive structures like this: https://stackblitz.com/edit/stackblitz-starters-9csxbu?file=src%2Fmain.ts

mmalerba commented on packages/common/src/directives/ng_template_outlet.ts:
```diff
@@ -62,8 +62,15 @@ export class NgTemplateOutlet<C = unknown> implements OnChanges {
+   * Injector to be used within the embedded view. A value of "inherit" can be used to indicate
+   * that the injector should be inherited from the template's location in the instantiated DOM.
+   */
+  @Input() public ngTemplateOutletInjector: Injector | 'inherit' | null = null;
```
I'm open to bike-shedding on what we call this option. `inherit` kind of made sense to me, but if I think about what I would call the default option, I guess it would have to be `inheritButFromTheTemplateInstead` which is maybe a sign that `inherit` is a bad name.

alxhub replied:
`injector: 'outlet'` maybe? Since we're saying "use the outlet's injector"?

mmalerba commented on packages/common/src/directives/ng_template_outlet.ts:
```diff
@@ -85,11 +92,23 @@ export class NgTemplateOutlet<C = unknown> implements OnChanges {
+   */
+  private _getInjector(): Injector | undefined {
+    if (this.ngTemplateOutletInjector === 'inherit') {
+      this._childInjector =
```
I assume its ok to cache the child injector the first time I create it?

alxhub replied:
(caching should be fine, but as below creating a separate one shouldn't be necessary)

mmalerba commented on packages/common/src/directives/ng_template_outlet.ts:
```diff
@@ -85,11 +92,23 @@ export class NgTemplateOutlet<C = unknown> implements OnChanges {
+  private _getInjector(): Injector | undefined {
+    if (this.ngTemplateOutletInjector === 'inherit') {
+      this._childInjector =
+        this._childInjector ?? Injector.create({providers: [], parent: this.injector});
```
I'm not really sure _why_ I need to create a child injector instead of just passing along the current injector, but based on my stackblitz testing it did seem to be necessary. Just wanted to call that out in case I'm doing something wrong

alxhub replied:
It shouldn't be necessary I don't think...

mmalerba replied:
Here's a modified version of my stackblitz to show that it really does seem to need me to create a child to get the same behavior as composing with content projection

mmalerba replied:
I talked to @crisbeto about this. He says this behavior makes sense to him because embedded view injectors can't be injected using `inject(Injector)`. By creating the child injector, we create an injector that acts as an embedded injector, but it ends up consulting the node injector for each lookup.

mmalerba reviewed (commented):
@alxhub couple questions for you to make sure I'm on the right track

mmalerba committed:
fix(core): correctly handle SkipSelf when resolving from embedded view injector

The `SkipSelf` flag is intended for the current injection context. When delegating to an embedded view injector, we are traversing to a parent/fallback scope, so the 'Self' has already been skipped.

mmalerba committed:
feat(common): add an 'outlet' injector option for ngTemplateOutlet

Adds an option (`ngTemplateOutletInjector="outlet"`) that instructs the ngTemplateOutlet to inherit its injector from the outlet's place in the instantiated DOM.

JeanMeche reviewed (approved):
reviewed-for: public-api

AndrewKushnir commented:
Caretaker note: [TGP](https://fusion2.corp.google.com/presubmit/845974920/OCL:845974920:BASE:866710630:1770452381289:51d88a89/targets) is "green".

AndrewKushnir commented:
This PR was merged into the repository. The changes were merged into the following branches:

- main: 18003a33bb0d6bb09def8a0e5939fa24069696eb
